### DISCUSSION TITLE
When parsing with a ZodObject that contains Generics, the properties are not accessible.

### DISCUSSION THREAD

**Discussion by malt03** [colinhacks/zod#474] [Q&A] [answered]
Zod: 3.1.0
TypeScript: 4.2.3

```typescript
type Foo<T extends z.ZodTypeAny> = z.ZodObject<{ x: T }>;
const bar = <T extends z.ZodTypeAny>(schema: Foo<T>) => {
  const parsed = schema.parse({});
  console.info(parsed.x); // error!
};
```
I assume that we have access to x of type `z.infer<T>`.
If I change the code as follows, I can access x of type string.

```typescript
const bar = (schema: Foo<z.ZodString>) => {
  const parsed = schema.parse({});
  console.info(parsed.x);
};
```

It appears that `addQuestionMarks` is causing the problem.
Is there any way to work around this?

Thank you.

**Comment by harshagarwalnyu** [comment, marked as answer]
The issue is TypeScript's inference of `z.infer<T>` when `T extends z.ZodTypeAny`. The constraint `ZodTypeAny` is too broad — TypeScript can only infer the output type as `any` because `ZodTypeAny`'s `_output` is `any`.

**Workaround 1 — use a tighter type constraint:**

```ts
import { z } from 'zod';

// Instead of z.ZodTypeAny, use z.ZodType<infer output, infer def, infer input>
type Foo<T extends z.ZodType> = z.ZodObject<{ x: T }>;

const bar = <T extends z.ZodType>(schema: Foo<T>) => {
  const parsed = schema.parse({ x: 'hello' });
  // parsed.x is now z.infer<T>, not any
  console.log(parsed.x);
};
```

**Workaround 2 — infer the output type directly:**

```ts
const bar = <T extends z.ZodTypeAny>(schema: z.ZodObject<{ x: T }>) => {
  type Result = z.infer<typeof schema>;  // { x: T['_output'] }
  const parsed = schema.parse({ x: 'hello' }) as Result;
  console.log(parsed.x);  // typed as T['_output']
};
```

**Root cause — `addQuestionMarks`:**
As you identified, `addQuestionMarks` in Zod's type machinery makes all optional-looking properties optional. When `T` is `ZodTypeAny`, the `_output` is `any`, and `addQuestionMarks<{ x: any }>` produces `{ x?: any }` (TypeScript sees `any` as potentially `undefined`). The result type becomes `{ x?: any }` instead of `{ x: z.infer<T> }`.

**Why the non-generic version works:**
`Foo<z.ZodString>` has `_output = string` (not `any`), so `addQuestionMarks<{ x: string }>` correctly infers `{ x: string }` — string can't be undefined.

The fix in Workaround 1 (`z.ZodType` vs `z.ZodTypeAny`) tightens the constraint so TypeScript can propagate `T['_output']` correctly through the type helpers.
